File: src/com/leetcode/Medium/LargestDivisibleSubset.java
package com.leetcode.Medium;

import java.util.*;

public class LargestDivisibleSubset {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        Arrays.sort(nums);
        int[] parent = new int[nums.length];
        int[] count = new int[nums.length];
        int max = 0, maxIndex = -1;
        for (int i = nums.length - 1; i >= 0; i--) {
            for (int j = i; j < nums.length; j++) {
                if (nums[j] % nums[i] == 0 && count[i] < 1 + count[j]) {
                    count[i] = 1 + count[j];
                    parent[i] = j;
                    if (count[i] > max) {
                        max = count[i];
                        maxIndex = i;
                    }
                }
            }
        }
        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < max; i++) {
            res.add(nums[maxIndex]);
            maxIndex = parent[maxIndex];
        }
        return res;
    }
}

File: JavaScript/Medium/LargestDivisibleSubset.js
/**
 * 368. Largest Divisible Subset
 * @param {number[]} nums
 * @return {number[]}
 */
const largestDivisibleSubset = function (nums) {
  nums.sort((a, b) => a - b);
  const parent = new Array(nums.length);
  const count = new Array(nums.length).fill(0);
  let max = 0, maxIndex = -1;

  for (let i = nums.length - 1; i >= 0; i--) {
    for (let j = i; j < nums.length; j++) {
      if (nums[j] % nums[i] === 0 && count[i] < 1 + count[j]) {
        count[i] = 1 + count[j];
        parent[i] = j;
        if (count[i] > max) {
          max = count[i];
          maxIndex = i;
        }
      }
    }
  }

  const res = [];
  for (let i = 0; i < max; i++) {
    res.push(nums[maxIndex]);
    maxIndex = parent[maxIndex];
  }
  return res;
};